[forward ad] Fix codegen to ignore undefined outputs

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81114
* #81055
* #79996

I don't think there's a way to avoid functions returning undefined tensors as outputs, so codegen will have to detect them before calling _set_fw_grad. Alternatively, we can just make calling _set_fw_grad with undefined self a no-op, but I'm biasing toward keeping _set_fw_grad more strict in case it is called in other areas.

Fixes https://github.com/pytorch/pytorch/issues/81111

soulitzer committed:
[forward ad] Fix codegen to ignore undefined outputs

[ghstack-poisoned]

albanD reviewed (commented):
Sounds good in principle. Needs testing.

soulitzer committed:
Update on "[forward ad] Fix codegen to ignore undefined outputs"


I don't think there's a way to avoid functions returning undefined tensors as outputs, so codegen will have to detect them before calling _set_fw_grad. Alternatively, we can just make calling _set_fw_grad with undefined self a no-op, but I'm biasing toward keeping _set_fw_grad more strict in case it is called in other areas.

Fixes https://github.com/pytorch/pytorch/issues/81111

[ghstack-poisoned]

soulitzer commented:
Testing added!

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1419,6 +1422,7 @@ def emit_fw_derivatives() -> List[str]:
                             out_arg=res[0], is_inplace=is_inplace_str
                         )
                     )
+                    requires_fw_grad += f" && ({derivative.var_names[0]}.defined())"
```
Don't we already check above that this is defined in https://github.com/pytorch/pytorch/blob/1afb804f26c960aec32aed7c587dcb483be8c580/torch/csrc/autograd/functions/utils.h#L100 ?

soulitzer replied:
Yeah that is only applied to the inputs of the jvp though

albanD reviewed (approved):
SGTM
nit question about double defined check

soulitzer commented:
@pytorchbot merge -g
